src/lyra/abstract_domains/data_structures/dict_segment_lattice.py

```diff
@@ -0,0 +1,298 @@
+"""
+Dictionary Segment Lattice
+========================
+
+Generic lattice to abstract dictionaries as a set of abstract segments
+
+:Authors: Lowis Engel
+"""
+
+from copy import deepcopy, copy
+from typing import List, Tuple, Set, Type, Dict, Any
+
+from lyra.abstract_domains.lattice import Lattice
+from lyra.core.expressions import VariableIdentifier
+from lyra.core.utils import copy_docstring
+
+
+class DictSegmentLattice(Lattice):
+    """Dictionary segment lattice element::
+
+        set of abstract segments,
+        represented by tuples (k,v) from two abstract domains K, V given as input
+        (which abstract a set of keys and their corresponding values),
+        s.t. (i) no two segments overlap (i.e. the meet of the keys is bottom)
+             (ii) there are no empty segments (i.e. neither k nor v are bottom)
+
+    Top: {(top, top)}
+    Bottom: {}
+
+    The default lattice element is Top, meaning the dictionary can contain anything.
+
+    .. document private methods
+    .. automethod:: DictSegmentLattice._less_equal
+    .. automethod:: DictSegmentLattice._meet
+    .. automethod:: DictSegmentLattice._join
+    .. automethod:: DictSegmentLattice._widening
+    """
+
+    # use typing.Generic instead of domains as parameter?
+    def __init__(self, key_domain: Type[Lattice], value_domain: Type[Lattice],
+                 key_d_args: Dict[str, Any] = {}, value_d_args: Dict[str, Any] = {},
+                 segments: Set[Tuple[Lattice, Lattice]] = None):
+        """
+        :param key_domain: domain for abstraction of dictionary keys,
+            ranges over the scalar variables and the special key variable v_k
+        :param value_domain: domain for abstraction of dictionary values,
+            ranges over the scalar variables and the special value variable v_v
+        :param key_d_args: arguments for the key_domain
+        :param value_d_args: arguments for the value_domain
+        :param segments: initial set of abstract segments; if None, a top segment is created
+        """
+        # TODO: segments internally as List?
+        super().__init__()
+        self._k_domain = key_domain
+        self._k_d_args = key_d_args
+        self._v_domain = value_domain
+        self._v_d_args = value_d_args
+
+        if segments is None:    # default element
```

Comment:
Rather than explicitly checking if segments is None, you can check if segments is _not_ None just with `if segments` (and swap the two branches):

```
if segments:
    self._segments = segments
else:
   self._segments = {(key_domain...
```